File: NiftyOptionCalculator-iOS/NiftyOptionCalculator/Models/TradeJournal.swift
import Foundation

// MARK: - Journal Tag

enum JournalTag: String, CaseIterable, Codable {
    case scalp
    case swing
    case hedging
    case spread
    case directional
    case expiry_day

    var displayName: String {
        rawValue.replacingOccurrences(of: "_", with: " ").capitalized
    }
}

// MARK: - Trade Mood

enum TradeMood: String, CaseIterable, Codable {
    case confident
    case nervous
    case neutral
    case greedy
    case fearful
    case disciplined

    var emoji: String {
        switch self {
        case .confident: return "\u{1F4AA}"   // flexed biceps
        case .nervous: return "\u{1F630}"     // anxious face
        case .neutral: return "\u{1F610}"     // neutral face
        case .greedy: return "\u{1F911}"      // money-mouth face
        case .fearful: return "\u{1F628}"     // fearful face
        case .disciplined: return "\u{1F9D8}" // person in lotus position
        }
    }
}

// MARK: - Market Condition

enum MarketCondition: String, CaseIterable, Codable {
    case trending_up
    case trending_down
    case range_bound
    case volatile
    case low_vol

    var displayName: String {
        rawValue.replacingOccurrences(of: "_", with: " ").capitalized
    }
}

// MARK: - Trade Outcome

enum TradeOutcome: String, CaseIterable, Codable {
    case profit
    case loss
    case breakeven
}

// MARK: - Journal Entry

struct JournalEntry: Identifiable, Codable {
    let id: String
    let user_id: String
    let symbol: String
    let strike_price: Double
    let option_type: String   // CE or PE
    let direction: String     // buy or sell
    let entry_price: Double
    let exit_price: Double?
    let quantity: Int
    let lot_size: Int
    let entry_date: String
    let exit_date: String?
    let expiry_date: String?
    let realized_pnl: Double?
    let realized_pnl_percent: Double?
    let title: String?
    let notes: String?
    let tags: String?
    let market_condition: String?
    let mood: String?
    let outcome: String?
    let paper_position_id: String?
    let created_at: String
    let updated_at: String
}

// MARK: - Journal List Response

struct JournalListResponse: Codable {
    let entries: [JournalEntry]
    let total: Int
}

// MARK: - Journal Stats

struct JournalStats: Codable {
    let total_trades: Int
    let winning_trades: Int
    let losing_trades: Int
    let breakeven_trades: Int
    let win_rate: Double
    let total_pnl: Double
    let average_pnl: Double
    let best_trade: Double?
    let worst_trade: Double?
    let average_holding_days: Double?
    let most_traded_symbol: String?
    let most_used_tags: [String]
    let trades_by_mood: [String: Int]
    let trades_by_condition: [String: Int]
}

// MARK: - Create Request

struct JournalCreateRequest: Codable {
    let symbol: String
    let strike_price: Double
    let option_type: String
    let direction: String
    let entry_price: Double
    let exit_price: Double?
    let quantity: Int
    let lot_size: Int
    let entry_date: String
    let exit_date: String?
    let expiry_date: String?
    let title: String?
    let notes: String?
    let tags: String?
    let market_condition: String?
    let mood: String?
    let outcome: String?
    let paper_position_id: String?
}

// MARK: - Update Request

struct JournalUpdateRequest: Codable {
    var exit_price: Double?
    var exit_date: String?
    var title: String?
    var notes: String?
    var tags: String?
    var market_condition: String?
    var mood: String?
    var outcome: String?
}

File: NiftyOptionCalculator-Android/app/src/main/java/com/niftyoption/calculator/data/models/TradeJournal.kt
package com.niftyoption.calculator.data.models

data class JournalEntry(
    val id: String = "",
    val user_id: String = "",
    val symbol: String,
    val strike_price: Double,
    val option_type: String,  // CE or PE
    val direction: String,    // buy or sell
    val entry_price: Double,
    val exit_price: Double? = null,
    val quantity: Int,
    val lot_size: Int,
    val entry_date: String,
    val exit_date: String? = null,
    val expiry_date: String? = null,
    val realized_pnl: Double? = null,
    val realized_pnl_percent: Double? = null,
    val title: String? = null,
    val notes: String? = null,
    val tags: String? = null,
    val market_condition: String? = null,
    val mood: String? = null,
    val outcome: String? = null,
    val paper_position_id: String? = null,
    val created_at: String = "",
    val updated_at: String = ""
)

data class JournalListResponse(
    val entries: List<JournalEntry>,
    val total: Int
)

data class JournalStats(
    val total_trades: Int,
    val winning_trades: Int,
    val losing_trades: Int,
    val breakeven_trades: Int,
    val win_rate: Double,
    val total_pnl: Double,
    val average_pnl: Double,
    val best_trade: Double? = null,
    val worst_trade: Double? = null,
    val average_holding_days: Double? = null,
    val most_traded_symbol: String? = null,
    val most_used_tags: List<String> = emptyList(),
    val trades_by_mood: Map<String, Int> = emptyMap(),
    val trades_by_condition: Map<String, Int> = emptyMap()
)

data class JournalCreateRequest(
    val symbol: String,
    val strike_price: Double,
    val option_type: String,
    val direction: String,
    val entry_price: Double,
    val exit_price: Double? = null,
    val quantity: Int,
    val lot_size: Int,
    val entry_date: String,
    val exit_date: String? = null,
    val expiry_date: String? = null,
    val title: String? = null,
    val notes: String? = null,
    val tags: String? = null,
    val market_condition: String? = null,
    val mood: String? = null,
    val outcome: String? = null
)

data class JournalUpdateRequest(
    val exit_price: Double? = null,
    val exit_date: String? = null,
    val title: String? = null,
    val notes: String? = null,
    val tags: String? = null,
    val market_condition: String? = null,
    val mood: String? = null,
    val outcome: String? = null
)

enum class JournalTag(val displayName: String) {
    SCALP("Scalp"),
    SWING("Swing"),
    HEDGING("Hedging"),
    SPREAD("Spread"),
    DIRECTIONAL("Directional"),
    EXPIRY_DAY("Expiry Day");

    val apiValue: String get() = name.lowercase()
}

enum class TradeMood(val displayName: String, val emoji: String) {
    CONFIDENT("Confident", "\uD83D\uDCAA"),
    NERVOUS("Nervous", "\uD83D\uDE30"),
    NEUTRAL("Neutral", "\uD83D\uDE10"),
    GREEDY("Greedy", "\uD83E\uDD11"),
    FEARFUL("Fearful", "\uD83D\uDE28"),
    DISCIPLINED("Disciplined", "\uD83E\uDDD8");

    val apiValue: String get() = name.lowercase()
}

enum class MarketCondition(val displayName: String) {
    TRENDING_UP("Trending Up"),
    TRENDING_DOWN("Trending Down"),
    RANGE_BOUND("Range Bound"),
    VOLATILE("Volatile"),
    LOW_VOL("Low Vol");

    val apiValue: String get() = name.lowercase()
}

enum class TradeOutcome(val displayName: String) {
    PROFIT("Profit"),
    LOSS("Loss"),
    BREAKEVEN("Breakeven");

    val apiValue: String get() = name.lowercase()
}
